kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -1,173 +1,146 @@
 import org.slf4j.Logger;
 
-/**
- * Coordinates communication between the StreamThread and execution threads, as
- * well as protecting and managing shared objects and accesses
- */
 public class AsyncThreadPool {
+  private final ExecutorService executorService;
+  private final Map<InFlightWorkKey, Map<AsyncEvent, Future<?>>> inFlight = new HashMap<>();
 
   public static final String ASYNC_THREAD_NAME = "AsyncThread";
 
-  private final LogContext logPrefix;
   private final Logger log;
-
-  private final String streamThreadName;
-
-  // TODO: start up new threads when existing ones die to maintain the target size
-  private final int threadPoolSize;
-  private final Map<String, AsyncThread> threadPool;
-  private final WriteOnlyProcessingQueue processingQueue;
+  private final AtomicInteger threadNameIndex = new AtomicInteger(0);
 
   public AsyncThreadPool(
       final String streamThreadName,
-      final int threadPoolSize
-  ) {
-    this.streamThreadName = streamThreadName;
-    this.threadPoolSize = threadPoolSize;
-    this.threadPool = new HashMap<>(threadPoolSize);
-
-    this.logPrefix = new LogContext(String.format("stream-thread [%s] ", streamThreadName));
+      final int size) {
+    final LogContext logPrefix
+        = new LogContext(String.format("stream-thread [%s] ", streamThreadName));
     this.log = logPrefix.logger(AsyncThreadPool.class);
-
-    final ProcessingQueue processingQueue = new ProcessingQueue(logPrefix);
-    this.processingQueue = processingQueue;
-    startThreads(processingQueue);
+    // todo: set the thread names
+    executorService = Executors.newFixedThreadPool(size, r -> {
+      final var t = new Thread(r);
+      t.setDaemon(true);
+      t.setName(
+          streamThreadName + "-" + ASYNC_THREAD_NAME + "-" + threadNameIndex.getAndIncrement());
+      return t;
+    });
   }
 
-  private void startThreads(final ProcessingQueue processingQueue) {
-    for (int asyncThreadIndex = 0; asyncThreadIndex < threadPoolSize; ++asyncThreadIndex) {
-      final String name = generateAsyncThreadName(streamThreadName, asyncThreadIndex);
-      final AsyncThread thread = new AsyncThread(name, processingQueue);
-
-      thread.setDaemon(true);
-      threadPool.put(name, thread);
-    }
+  @VisibleForTesting
+  Map<AsyncEvent, Future<?>> getInFlight(final String processorName, final int partition) {
+    return inFlight.get(InFlightWorkKey.of(processorName, partition));
+  }
 
-    for (final AsyncThread thread : threadPool.values()) {
-      thread.start();
+  public void removeProcessor(final String processorName, final int partition) {
+    log.info("clean up records for {}:{}", processorName, partition);
+    final var key = InFlightWorkKey.of(processorName, partition);
+    final Map<AsyncEvent, Future<?>> inFlightForTask = inFlight.remove(key);
+    if (inFlightForTask != null) {
+      inFlightForTask.values().forEach(f -> f.cancel(true));
```

Comment:
I guess this is another reason to keep the total number of in-flight events pretty low. Closing processors should be fast. But I guess there's no problem since we'll add a config to make sure this isn't a huge list